src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs

```diff
@@ -79,10 +79,13 @@ public AcquireTokenForClientParameterBuilder WithSendX5C(bool withSendX5C)
         /// By default the value for this variable is false.
         /// See https://aka.ms/msal-net-region-discovery for more details.
         /// </summary>
-        /// <param name="autoDetectRegion"><c>true</c> if the token request should be sent to regional ESTS. Otherwise <c>false</c>.
-        /// The default is <c>false</c></param>
+        /// <param name="autoDetectRegion"><c>true</c> if the token request should be sent to regional ESTS. The default is <c>false</c>.
+        /// </param>
+        /// <param name="useRegion"> optional parameter to provide region to MSAL. This parameter will be used along with auto detection of region.
+        /// If the region is auto detected, the provided region will be compared with the detected region and used in telemetry to do analysis on correctness of the region provided.
+        /// If auto region detection fails, the provided region will be used for instance metadata.</param>
```

Comment:
The auto detect region is set to true if the user wants to opt for regional endpoint. If set to false or not set, the requests will go to the global endpoint as usual.
UseRegion is an optional api where a user can provide the region in case they have it set in their config or they know the region. In case we are able to detect the region we will use the auto detected region and compare it with the region provided by the user to analyse the correctness of the region provided. In case we fail to detect the region and we have a value in useRegion, we will go ahead and use this region to hit the regional endpoint. And if in case user provided some garbage value in that place and we form a regional endpoint using it, ests is working on a way to handle those cases.

Reply:
Why not just use the region, if it's provided, and skip the auto detect region call? Then, if the the call fails, I guess try to autoDetect the region and try the request again? Is this specific guidance from eSTS on how to handle this?

---
In reply to: [549887878](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2315#discussion_r549887878) [](ancestors = 549887878)

Reply:
This is required to analyze the validity of the value that user provides. We know that there are azure services where we cannot detect the region and there can be more platforms where auto detection fails. For this reason it was decided to allow the user to provide the region. But this is something we cannot completely rely on as it could be hardcoded or the value of region they provide might not be the actual azure region. So, we will analyze the validity of the region provided by the user and in the initial phases of adoption we can let them know that there is a mismatch in the region they provide and region that was detected. Also, in case we fail to compare we are logging the region and region source which can help to ensure the region provided is correct and the calls are going to regional endpoint. If user opts to use regional endpoint goal is to hit the regional endpoint and use global only as a last option.

---
In reply to: [551511449](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2315#discussion_r551511449) [](ancestors = 551511449,549887878)

Reply:
Thanks for the explanation, @neha-bhargava, I had a similar question as @jennyf19 

In that case, what about having the following public API: `WithAzureRegion(string regionToUse="autoDetect")`:
- If the developer provides the region, then for sure the boolean would be true ? 
- and if they don't provide it (then the region would become) "autoDetect", which you'd then internally set boolean `autoDetectRegion` to true (but that would not be needed to be surfaced to the developers)

Note: you don't need to change your current method. Just make it internal, and wrap it into the API I propose?

I was also wondering if there is a value of passing `WithAzureRegion(false)`? What is the use case? would it be that the developer used the regional endpoint and then want to go to Global? but given this is on the AcquireTokenForClientParameterBuilder, the developer could just omit the `.WithAzureRegion()` to do that?

Reply:
Side note: We could just check for a default empty string, `WithAzureRegion(string regionToUse="")`. Would perhaps changing the parameter name to `fallbackRegion` be clearer?

Reply:
@jmprieur : the reason to have the boolean is to avoid developers having to write if statements e.g.
if (boolUseRegion)
  builder.withRegion(...)

instead they can write withyRegion(boolUseRegion, "valueFromMyConfig")

#CleanerCode

Reply:
@henrik-me how would that lead to not using if statements? if they don't want to use the region, then they just won't use .withRegion(). in the AT call, right?